src/database/api/DataSnapshot.ts

```diff
@@ -0,0 +1,177 @@
+import { validateArgCount, validateCallback } from '../../utils/validation';
+import { validatePathString } from '../core/util/validation';
+import { Path } from '../core/util/Path';
+import { PRIORITY_INDEX } from '../core/snap/indexes/PriorityIndex';
+import { Node } from '../core/snap/Node';
+import { Reference } from './Reference';
+import { Index } from '../core/snap/indexes/Index';
+import { ChildrenNode } from '../core/snap/ChildrenNode';
+
+/**
+ * Class representing a firebase data snapshot.  It wraps a SnapshotNode and
+ * surfaces the public methods (val, forEach, etc.) we want to expose.
+ */
+export class DataSnapshot {
+  /**
+   * @param {!Node} node_ A SnapshotNode to wrap.
+   * @param {!Reference} ref_ The ref of the location this snapshot came from.
+   * @param {!Index} index_ The iteration order for this snapshot
+   */
+  constructor(private readonly node_: Node,
+              private readonly ref_: Reference,
+              private readonly index_: Index) {
+  }
+
+  /**
+   * Retrieves the snapshot contents as JSON.  Returns null if the snapshot is
+   * empty.
+   *
+   * @return {*} JSON representation of the DataSnapshot contents, or null if empty.
+   */
+  val(): any {
+    validateArgCount('DataSnapshot.val', 0, 0, arguments.length);
+    return this.node_.val();
+  }
+
+  /**
+   * Returns the snapshot contents as JSON, including priorities of node.  Suitable for exporting
+   * the entire node contents.
+   * @return {*} JSON representation of the DataSnapshot contents, or null if empty.
+   */
+  exportVal(): any {
+    validateArgCount('DataSnapshot.exportVal', 0, 0, arguments.length);
+    return this.node_.val(true);
+  }
+
+  // Do not create public documentation. This is intended to make JSON serialization work but is otherwise unnecessary
+  // for end-users
+  toJSON(): any {
+    // Optional spacer argument is unnecessary because we're depending on recursion rather than stringifying the content
+    validateArgCount('DataSnapshot.toJSON', 0, 1, arguments.length);
+    return this.exportVal();
+  }
+
+  /**
+   * Returns whether the snapshot contains a non-null value.
+   *
+   * @return {boolean} Whether the snapshot contains a non-null value, or is empty.
```

Comment:
Just chiming in since I was thinking about this the other day. Type annotations in the comments only make sense when using a tool like Closure compiler to do type checking, but it's not necessary anymore with TypeScript. Besides, I bet the types in the comments won't always be updated when the code changes so eventually it might do more harm than good.